cmd/autoscale-scheduler/main.go

```diff
@@ -12,9 +15,30 @@ import (
 // all of the juicy bits are defined in pkg/plugin/
 
 func main() {
-	command := app.NewSchedulerCommand(app.WithPlugin(plugin.Name, plugin.NewAutoscaleEnforcerPlugin))
-	if err := command.Execute(); err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+	if err := runProgram(); err != nil {
+		log.Fatal(err)
 	}
 }
+
+// runProgram is the "real" main, but returning an error means that
+// the shutdown handling code doesn't have to call os.Exit, even indirectly.
+func runProgram() error {
+	// this: listens for sigterm, when we catch that signal, the
+	// context gets canceled, a go routine waits for half a second, and
+	// then closes the signal channel, which we block on in a
+	// defer. because defers execute in LIFO errors, this just
+	// pauses for a *very* short period of time before exiting.
+	//
+	// eventually, the constructed application
+	sig := make(chan struct{})
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGTERM)
+	go func() { defer close(sig); <-ctx.Done(); time.Sleep(500 * time.Millisecond) }()
+	defer func() { <-sig }()
+	defer cancel()
+
+	command := app.NewSchedulerCommand(app.WithPlugin(plugin.Name, plugin.NewAutoscaleEnforcerPlugin(ctx)))
```

Comment:
I actually think that by capturing the context here which wraps the other parts of the scheduler's shut down, we get our own orderly shutdown regardless of kube's 

(modulo actually waiting for things to shut down, I believe that we will have correctly passed a context through to all of the parts that could use one (the ones that don't get this one, will get one from the scheduler which has a similar trigger.) 

Also, unfortunately, kube doesn't use the context we pass to cobra, which is... unfortunate, but we tried.)

Reply:
That _might_ be the case. It's hard to say without looking into why the current shutdowns are stalling (e.g., maybe the scheduler waits for all plugin goroutines to exit before returning, so just waiting for `command.ExecuteContext` to finish might block).

Reply:
yeah, I definitely want to debug this. more, but I think resolving the long shutdowns should be debugged on its own.